budget/app.py

```diff
@@ -106,8 +137,93 @@ def compare_budgets_and_users(account_id, users):
 
 def create_budget(synapse_id, team):
   '''Creates an AWS budget for a synapse user id'''
-  # TODO implement
-  log.info('Placeholder -- this will make a budget')
+  log.debug(f'Creating budget for synapse user {synapse_id}, member of team {team}')
+
+  team_budget_rules = TEAM_BUDGET_RULES.get('teams').get(team)
+  if not team_budget_rules:
+    raise ValueError(f'No budget rules available for team {team}')
+  budget_amount = team_budget_rules['amount']
+  budget_period = team_budget_rules['period']
+  budget = {
+    'BudgetName': f'{BUDGET_NAME_PREFIX}{synapse_id}',
+    'BudgetLimit': {
+      'Amount': budget_amount,
+      'Unit': 'USD'
+    },
+    'CostFilters': {
+      'TagKeyValue': [
+        (
+          'aws:servicecatalog:provisioningPrincipalArn$arn:aws:sts::'
+          f'{_get_account_id()}:assumed-role/ServiceCatalogEndusers/{synapse_id}'
+        )
+      ]
+    },
+    'CostTypes': {
+      'IncludeRefund': False,
+      'IncludeCredit': False
+    },
+    'TimeUnit': budget_period,
+    'BudgetType': 'COST'
+  }
+  budgets_client = get_client('budgets')
+  return budgets_client.create_budget(
+    AccountId=_get_account_id(),
+    Budget=budget
+    )
+
+
+def _create_budget_notification(synapse_id, threshold, admin_emails=None):
+  '''Creates a notification rule when the budget crosses a particular threshold.
+
+  May have multiple recipients. User notifications are sent through SNS
+  in order to send email to Synapse addresses, which cannot receive email
+  unless sent through the Synapse system, but emails to administrators can
+  be sent directly.
+  '''
+  budgets_client = get_client('budgets')
+  budget_name = f'{BUDGET_NAME_PREFIX}{synapse_id}'
+  subscribers = []
+  if admin_emails:
+    subscribers = [{
+      'SubscriptionType': 'EMAIL',
+      'Address': address
+    } for address in admin_emails]
+  subscribers.append({
+    'SubscriptionType': 'SNS',
```

Comment:
hmm, SNS email notifications require confirmation by the end user. it looks like this...
```
You have chosen to subscribe to the topic:
arn:aws:sns:us-east-1:044455566678:agora-prod-notifications-SnsTopic-14EDD9ZWD8A8P

To confirm this subscription, click or visit the link below (If this was in error no action is necessary):
Confirm subscription
```
Which also means they will get this subscription request every time they are added to a team or change teams.
 
I wonder if we can bypassing confirmation?

Reply:
I believe @brucehoff will need to solve that in his [PR](https://github.com/Sage-Bionetworks-IT/lambda-send-budget-alert/pull/1).